### ISSUE TITLE
Alertmanager not disinhibiting alert

### ISSUE THREAD

**Issue by oscarruizmrf** [prometheus/alertmanager#1153]
**What did you do?**
We configured a couple of alarms and expect one of them (tooManyAlarms) to inhibit the other one (lowRevenueComparedToPredction).

**What did you expect to see?**
In the first 10 minutes range, alert tooManyAlarms inhibits lowRevenueComparedToPrediction that fires every 2 minutes. Alertmanager shows lowRevenueComparedToPrediction inhibited, as expected. Also, there is no notification via Slack about lowRevenueComparedToPrediction alarms, as expected.

But, in the next 10 minutes range, tooManyAlarms is not firing, hence I expect lowRevenueComparedToPrediction to fire every 2 minutes and notify via Slack.
I expect to see the alarm (lowRevenueComparedToPrediction) in alarmmanager not inhibited. 

**What did you see instead? Under which circumstances?**
 Alertmanager shows lowRevenueComparedToPrediction firing but inhibited although tooManyAlarms is not firing. So, once lowRevenueComparedToPrediction is inhibited, it never disinhibits, unless we restart alertmanager. In that case, lowRevenueComparedToPrediction fires, notifies via Slack and appears not inhibited in alertmanager, until next inhibition.

**Environment**

* System information:

	Darwin 17.2.0 x86_64

* Alertmanager version:

Alertmanager is not showing version, got it from github "github.com/prometheus/alertmanager/cmd/..."
alertmanager, version  (branch: , revision: )
  build user:       
  build date:       
  go version:       go1.9.2

* Prometheus version:

Not showing version on --version.
prometheus, version  (branch: , revision: )
  build user:       
  build date:       
  go version:       go1.9.2

* Alertmanager configuration file:
```
route:
  # Whether an alert should continue matching subsequent sibling nodes.
  receiver: 'slack_team'
  group_by: ['alertname']
  group_wait: 30s
  group_interval: 15s
  repeat_interval: 30m
  routes:
  - match_re:
      jiraProject: ^([a-bA-Z]+)$
    receiver: 'slack_team'
  - match:
      owner: test
    receiver: 'slack_test'
    continue: true  

inhibit_rules:
- target_match_re:
    severity: warning
    group: .+  
  source_match_re:
    severity: critical
    group: .+
  # Apply inhibition if the group is the same.
  equal: ['group']
```

* Prometheus configuration file:
```
Alerts for replicating the issue.
groups:
- name: Revenue
  interval: 10s
  rules:
  - alert: tooManyAlarms
    expr: minute() % 20 < 10
    labels:
      group: revenue
      severity: critical
      owner: test
    annotations:
      summary: "Fires every 10 minutes"
      description: "Fires every 10 minutes"
  - alert: lowRevenueComparedToPrediction
    expr: minute()%2 == 0
    labels:
      group: revenue
      severity: warning
      jiraProject: RV
    annotations:
      summary: "Low revenue compared to estimation"
      description: "Low revenue compared to estimation"   
```

* Logs:
```
insert Prometheus and Alertmanager logs relevant to the issue here
```

**Comment by beorn7** [comment]
The binaries don't show the version most likely because you have built them directly with the `go` command (rather than `make`). It would be good if you could provide the git sha1 where you built the binaries from (run `git rev-parse HEAD` in the respective git repo).

**Comment by oscarruizmrf** [comment]
Yes, I built with go, but this release, downloaded as binaries fails also.

alertmanager, version 0.10.0 (branch: HEAD, revision: 133c888ef3644b47a52acbaeffb09f4cc637df1b)
 build user:       root@01302b7cd08a
 build date:       20171109-15:36:24
 go version:       go1.9.2

**Comment by oscarruizmrf** [comment]
Also tested with latest alertmanager

alertmanager, version 0.12.0 (branch: HEAD, revision: fc33cc78036f82ef8d4734c197a96f7cb6c952a3)
  build user:       root@c9169eb10d06
  build date:       20171215-14:14:51
  go version:       go1.9.2

**Comment by beorn7** [comment]
Doesn't look like a recent regression then, thanks.

@stuartnelson3 do you have an idea what could be going on?

**Comment by beorn7** [comment]
This might have to do with the specific interaction of `group_wait`, `group_interval`, and `repeat_interval`. Could you try out what happens if you set them all to 15s?

**Comment by oscarruizmrf** [comment]
Hi, we set all to 15s, but there is no difference in the behaviour.
The alarm that should dishinibit, does not.

**Comment by oscarruizmrf** [comment]
Hi, any idea on this @beorn7 ?
Thanks.

**Comment by beorn7** [comment]
(Just back from vacations.)

I'll investigate this case in our setup. So far, we haven't run into the issue in our production environment despite using quite a bit of inhibitions (which made me think it might have to do with some unlucky timing of your various config settings, but setting everything to 15s in the config should rule that out).

But @stuartnelson3 and perhaps @brancz are much closer to the codebase than I am right now, so they might have ideas, too.

**Comment by oscarruizmrf** [comment]
@beorn7 can you provide us with some working config on inhibitions? it may help checking out differences to find the issue.

**Comment by beorn7** [comment]
There is really nothing magic with our inhibition rules. It's well possible that we just haven't seen this issue yet because the inhibited alert also ceased firing when the inhibiting alert did so. (The usual use case is to inhibit a bunch of alerts if a known common root cause is in effect that another team should be alerted about. Once that cause is removed, also the formerly inhibited alerts cease firing.)

Will investigate as soon as I find time.

**Comment by oscarruizmrf** [comment]
So, in your cases, when the inhibitor ceases, all inhibited cease as well?
What you are saying is that there is no disinhibition implemented?
Inhibited alerts disinhibit because they resolve, not because parent
resolves.

**Comment by oscarruizmrf** [comment]
Hi,
@beorn7, @stuartnelson3 or @brancz, if you give us some indications we can work on it to fix the issue. Where should we start looking?
Thanks.

**Comment by brancz** [comment]
Pretty much all code for inhibition is here: https://github.com/prometheus/alertmanager/blob/master/inhibit/inhibit.go

My gut feeling tells me we need to look at an "unmute" path, here: https://github.com/prometheus/alertmanager/blob/master/inhibit/inhibit.go#L141-L155

**Comment by fabxc** [comment]
I unsuccessfully tried to reproduce this. There is no inhibition state being cached or anything so it's also hard to think what it could be.

The only way is really that the inhibiting alert actually lingers around and isn't resolved. But then it should also show up in the UI.

Note that the timing you provided here (2 and 10 minutes) is well within a typical group interval range. Some delays are to be exepcted there. Your group interval is admittedly on the very short end of the range.
Did the inhibited alert eventually start firing again or did this unexpected behavior continue forever?

**Comment by oscarruizmrf** [comment]
Hi,
the inhibited alert never fires, as far as we have tested, for a couple of
hours, even when the source alert resolves.

**Comment by dmitrypapka1** [comment]
Will this be included in v0.15.0?

**Comment by simonpasquier** [comment]
@dmitrypapka1 yes #1331 has been merged to master so it will part of the future v0.15.0 release.
